### ISSUE TITLE
azurerm_network_watcher_flow_log creation fails if NSG+RG names > 63 characters

### ISSUE THREAD

**Issue by larslind** [terraform-providers/terraform-provider-azurerm#12460]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

$ ./terraform -v
Terraform v1.0.1
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.65.0

### Affected Resource(s)

* `azurerm_network_watcher_flow_log`

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
    }
  }
}

provider "azurerm" {
  features {}
}

variable "storage_account_name" {
  type    = string
}

variable "location" {
  type    = string
}

resource "azurerm_resource_group" "test" {
             #0        1         2
  name     = "rg345678901234567890"
  location = var.location
}

resource "azurerm_network_security_group" "test" {
                        #         1         2         3         4         5 
  name                = "nsg45678901234567890123456789012345678901234567890"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_network_watcher" "test" {
  name                = "acctestnw"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_storage_account" "test" {
  name                = var.storage_account_name
  resource_group_name = azurerm_resource_group.test.name
  location            = azurerm_resource_group.test.location

  account_tier              = "Standard"
  account_kind              = "StorageV2"
  account_replication_type  = "LRS"
  enable_https_traffic_only = true
}

resource "azurerm_log_analytics_workspace" "test" {
  name                = "acctestlaw"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku                 = "PerGB2018"
}


resource "azurerm_network_watcher_flow_log" "test" {
  network_watcher_name = azurerm_network_watcher.test.name
  resource_group_name  = azurerm_resource_group.test.name

  network_security_group_id = azurerm_network_security_group.test.id
  storage_account_id        = azurerm_storage_account.test.id
  enabled                   = true

  retention_policy {
    enabled = true
    days    = 7
  }

  traffic_analytics {
    enabled               = true
    workspace_id          = azurerm_log_analytics_workspace.test.workspace_id
    workspace_region      = azurerm_log_analytics_workspace.test.location
    workspace_resource_id = azurerm_log_analytics_workspace.test.id
    interval_in_minutes   = 10
  }
}
```

### Debug Output

```
azurerm_network_watcher_flow_log.test: Creating...
╷
│ Error: Error creating "Flow Log: (Name \"Microsoft.Networkrg345678901234567890nsg45678901234567890123456789012345678901234567890\" / Network Watcher Name \"acctestnw\" / Resource Group \"rg345678901234567890\")": network.FlowLogsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- OriginalError: Code="InvalidResourceName" Message="Resource name Microsoft.Networkrg345678901234567890nsg45678901234567890123456789012345678901234567890 is invalid. The name can be up to 80 characters long. It must begin with a word character, and it must end with a word character or with '_'. The name may contain word characters or '.', '-', '_'." Details=[]
│
│   with azurerm_network_watcher_flow_log.test,
│   on main.tf line 59, in resource "azurerm_network_watcher_flow_log" "test":
│   59: resource "azurerm_network_watcher_flow_log" "test" {
```

### Expected Behaviour

Either:

* Fix the flow log name generation formula to truncate the name to max 80 characters (or fewer) so that the `terraform apply` doesn't fail.

or

* Add a `name` argument to align with the other programmatic interfaces for deploying resources in Azure. As seen from the documentation extracts below, these don't generate a flow log name but take the name as a required input argument/parameter. This would be the "proper" solution, but would of course break all existing Terraform code that deploys flow logs.

[REST API](https://docs.microsoft.com/en-us/rest/api/network-watcher/flow-logs/create-or-update#flowlog): 
| Name | In | Required | Type | Description
|---|---|---|---|---|
| flowLogName | path | True | String | The name of the flow log. |

[ARM Template](https://docs.microsoft.com/en-us/azure/templates/microsoft.network/networkwatchers/flowlogs?tabs=json):

| Name | Type | Required | Value 
|---|---|---|---|
| name | string | yes | The name of the flow log. 

[Azure CLI](https://docs.microsoft.com/en-us/cli/azure/network/watcher/flow-log?view=azure-cli-latest#az_network_watcher_flow_log_create):

>#### **Required Parameters**
> ...
> 
>**--name -n**
>
>The name of the flow logger.
>
> ...

### Actual Behaviour

It should be possible to create a Network Watcher Flow Log even if the combined length of the Resource Group and Network Security Group names exceeds 80 characters. The current Terraform limit is actually 63 and if the combined length of the NSG+RG names exceeds this, the `terraform apply` fails.

The Terraform Resource Provider tries to emulate the behaviour of the Azure Portal, which generates the flow log name by concatenating the name of the NSG with the RG and then appending "-flowlog" at the end.

```
<NSG Name> + <Resource Group Name> + "-flowlog"
```

 Because the length of the flow log name is limited to 80 characters, the portal truncates the NSG+RG name string. In fact the total length of the flow log name generated by the portal is trucated to be at most 64 characters.

Below is an example with a 20-character RG name and a 40-character NSG name, which works with Terraform.

```
Ruler (10s):          0        1         2         3         4         5         6         7         8
Ruler:                12345678901234567890123456789012345678901234567890123456789012345678901234567890
RG Name:              rg345678901234567890
NSG Name:             nsg4567890123456789012345678901234567890
TF flow log name:     Microsoft.Networkrg345678901234567890nsg4567890123456789012345678901234567890
Portal flow log name: nsg4567890123456789012345678901234567890-rg3456789012345-flowlog
```
As can be seen above, the flow log name generated by Terraform takes the form 
```
"Microsoft.Network" + <Resource Group Name> + <NSG Name>
```
which is completely different from what the portal does.
### Steps to Reproduce

1. `terraform apply`

**Comment by larslind** [comment]
Thanks for tagging, @jackofallops, but given that with completely valid Resource Group and NSG names, the `terraform apply` fails, should this not be tagged as a `bug` rather than `enhancement`? It's currently a blocking issue for a major enterprise deployment.

**Comment by larslind** [comment]
And the bug is in line 43 of this piece of code: https://github.com/terraform-providers/terraform-provider-azurerm/blob/a3e73e9dc2b044eee8259d5aac7f712052fdafdb/azurerm/internal/services/network/parse/flow_log.go. 
CC @magodo

**Comment by magodo** [comment]
@larslind Thank you for submitting this issue :+1:

That said in the description of PR #11670 that introduced this:

> This refactors the `azurerm_network_watcher_flow_log` to using its own separate endpoint - rather than using the `configureFlowLog` endpoint of the Network Watcher (more details [here](https://github.com/terraform-providers/terraform-provider-azurerm/issues/11587#issuecomment-836303145))
>
> This PR is implemented in a way to avoid breaking changes on the existing instances and ensure backward compatiblility (locally tested).

Also, the [comment that above the line you mentioned](https://github.com/terraform-providers/terraform-provider-azurerm/blob/a3e73e9dc2b044eee8259d5aac7f712052fdafdb/azurerm/internal/services/network/parse/flow_log.go#L40-L42) above:

> The flow log name generated by the "configureFlowLog" endpoint is in below format: Microsoft.Network{nsg rg name}{nsg name}.  We follow this rule to ensure backward compatibility.

This means that if you create a network watcher flow log using the provider prior to (including) v2.62.0, then `GET` it via the network watcher flow log dedicated endpoint, you'll see the name of it follows the format I listed above.

So with your example above provisioned via v2.62.0 provider, and `GET` back via the dedicated endpoint, the response is like below:

```json
 {
            "name": "Microsoft.Networkrg345678901234567890nsg45678901234567890123456789012345678901234567890",
            "id": "/subscriptions/****/resourceGroups/rg345678901234567890/providers/Microsoft.Network/networkWatchers/acctestnw/flowLogs/Microsoft.Networkrg345678901234567890nsg45678901234567890123456789012345678901234567890",
            ...
  }
```

Here the name is of length:

```bash
💤 wc <<<"Microsoft.Networkrg345678901234567890nsg45678901234567890123456789012345678901234567890"
 1  1 88
```

(this apparently exceeds the API limit)

There is no hard requirement that we have to follow the portal behavior in Terraform. What we want to do instead is to guarantee no breaking changes in the same major version.

To fix this issue, I'd prefer to adding the `name` property as you have suggested in v3.0, and also adding the validation for the name length. WDYT?

**Comment by larslind** [comment]
Thanks @magodo . Adding `name` would make it consistent with the REST API, ARM template, and Azure CLI. If you want it to be compatible with existing Terraform code it would have to be an optional attribute and if it's not provided, the name calculated as above, but truncated to 80 characters.

**Comment by DotNetSparky** [comment]
@magodo is this something that's going to be fixed in the next update (referring to the error caused by >80 char names) -- I'll admit, I'm concerned about the 'v3.0.0' tag.

**Comment by magodo** [comment]
@DotNetSparky Unfortunately, yes.

The reason is that we have to keep backward compatibility, which means we will keep the way how a `azurerm_network_watcher_flow_log` is imported. Currently, it was imported by specifying the ID in form of: `<network_watcher_id>/networkSecurityGroupId/<nsg_id>`. During importing, the code will based on the ID construct its name (using the format I mentioned above) and read it. If we expose a `name` and users set it to be some other value, then the import will not work.

**Comment by larslind** [comment]
@magodo - would it be possible to do an interim release that just truncates the generated name to 80 characters and then add a new `name` parameter in v3? That would at least fix the current issue that `terraform apply` fails when the RG and NSG names are too long. This is a blocking issue for a major enterprise customer - ping me in Teams if you want further details.

**Comment by magodo** [comment]
@larslind I've drafted a PR for this.

While I'm still concerned with two following points:

- If the name pattern of the resource group name or the NSG name doesn't conform to the flow log pattern, there is nothing can be done in the provider code to work around that
- Before the refactoring in #11670, the `configureFlowLog` endpoint is used to create the flow log. The created flow log CAN has name longer than 80 in length. This means if we merge this PR, it will break the users who created a long name flow log prior to #11670, and wants to use the latest provider to import that resource.

**Comment by larslind** [comment]
> * [...] The created flow log CAN has name longer than 80 in length. This means if we merge this PR, it will break the users who created a long name flow log prior to [Refactor `azurerm_network_watcher_flow_log` and add supports for `location` and `tags` #11670](https://github.com/terraform-providers/terraform-provider-azurerm/pull/11670), and wants to use the latest provider to import that resource.

I understand, but since the latest provider can't create flog logs with names longer than 80 characters, then importing a resource with a name that is too long would be pointless anyway. I'm not sure what the best solution is for this - maybe a warning/error on import?

**Comment by DotNetSparky** [comment]
Just asking for clarification -- for those of us who have existing flowlogs (created by Terraform) with too-long of names, is this resolved in the latest provider (2.68.0) so that we can continue 'business as usual', or do we have to stay on v2.62.1 until v3 comes out?

(and prefacing my next Q with "I know, I'm asking that which should never be asked :)") -- If it is a v3 scenerio, is there any kind of really-rough-ETA? (e.g. could be in the next month, or likely not until end of year or next year).  I ask only because if it's not something soon, then I can start a discussion amongst our teams about making an exception to our naming policy for NSGs to try to get the generated names to be shorter in length.

**Comment by DotNetSparky** [comment]
I was able to answer my own question, actually--the 'truncate' fix looks like it was merged in for the v2.68.0 release (I didn't see it mentioned in the release notes).

I just did a 'terraform plan' and it wants to create some flowlogs that already exist (which is understandable, because TF is now looking for a truncated name)--I can live with that, I wrote a script to drop all existing flowlogs and then I can let Terraform recreate them all (easier/quicker in my case than doing them one-by-one).

(leaving my comments here in case other people weren't sure if it was safe to update AzureRM)
